File: isosplit/isosplit1d.m
function [labels,pp]=isosplit1d(X,opts)

if (nargin<1) test_isosplit1d; return; end;
if (isstr(X))
	if (strcmp(X,'calibrate'))
		do_calibration(opts);
	else
		error('Invalid string command.');
	end;
	return;
end;

if (nargin<2) opts=struct; end;
opts.minsize=4; %hard coded to be consistent with calibration.

N=length(X);

%read the calibration (speed this up?)
tic1=tic;
[avg,stdev,scores]=read_isosplit_calibration(N);
if (isfield(opts,'m_max'))
	avg=avg(1:opts.m_max);
	stdev=stdev(1:opts.m_max);
end;
curve_len=length(avg);
num_trials=length(scores);
if (length(avg)==0) 
	error(sprintf('isosplit1d: calibration not performed for N=%d\n',N));
end;
%fprintf('tic1: %f\n',toc(tic1));

%compute the curve
tic2=tic;
[curve,cutpoint]=get_isosplit_curve(X,curve_len,opts);
%fprintf('tic2: %f\n',toc(tic2));

tic3=tic;
%compute the score by comparing to calibration avg and stdev
% and then compute the pp
inds=find(stdev~=0);
if (length(inds)>0)
	diff0=(avg(inds)-curve(inds)');
	%we need a change of at least 0.1 to count it
	%this is important in the case where stdev is extremely close to 0
	diff0(abs(diff0)<0.1)=0;
	score0=max(diff0./stdev(inds));
	pp=length(find(scores<score0))/num_trials;
else
	score0=0;
	pp=0;
end;
%fprintf('tic3: %f\n',toc(tic3));

tic4=tic;
labels=zeros(1,length(X));
labels(find(X<cutpoint))=1;
labels(find(X>=cutpoint))=2;
%fprintf('tic4: %f\n',toc(tic4));

end

function fname=get_calibration_fname(N)

dirname=fileparts(mfilename('fullpath'));
fname=sprintf('%s/isosplit_calibration/isosplit_calibration_%d.txt',dirname,N);

end


function [curve,cutpoint]=get_isosplit_curve(X,curve_len,opts)
	N=length(X);
	if (N<opts.minsize*2)
		curve=zeros(1,curve_len);
		cutpoint=0;
		return;
	end;
	X=sort(X);
	spacings=[X(2)-X(1),(X(3:end)-X(1:end-2))/2,X(end)-X(end-1)];
	density=1./spacings;
	logdensity=log(density);
    tA=tic;
	fit1=jisotonic(logdensity,'updown');
    fprintf('elapsed for updown: %.3f\n',toc(tA));
	resid1=logdensity-fit1;
	%fit2=jisotonic(resid1,'downup');
	%fit2=jisotonic(resid1,'downup',spacings);
    tA=tic;
	fit2=jisotonic(resid1,'downup',sqrt(spacings)); %this weighting seems to work best
    fprintf('elapsed for downup: %.3f\n',toc(tA));
	fit2=fit2-mean(fit2); %added 6/15/15, not sure if necessary
	%fit2=jisotonic(resid1,'downup');
	fit2_sorted=sort(fit2(opts.minsize:end-opts.minsize));
	curve=cumsum(fit2_sorted)./(1:length(fit2_sorted));
	if (length(curve)<curve_len)
		curve=[curve,zeros(1,curve_len-length(curve))];
	end;
	curve=curve(1:curve_len);
	[~,ind]=min(fit2(opts.minsize:end-opts.minsize));
	ind=ind+opts.minsize-1;
	if (ind<=1) cutpoint=X(1); return; end;
	if (ind>=length(X)) cutpoint=X(end); return; end;
	if (spacings(ind-1)>spacings(ind+1))
		cutpoint=(X(ind-1)+X(ind))/2;
	else
		cutpoint=(X(ind)+X(ind+1))/2;
	end;
end


function do_calibration(start_kk)

dirname=fileparts(get_calibration_fname(1));
if (~exist(dirname,'dir'))
	mkdir(dirname);
end;

curve_len=1000;
num_trials=1000;
incr=1.02;

kk=start_kk;
while true
	N=floor(incr^kk);
	fname=get_calibration_fname(N);
	if (~(exist(fname,'file')==2))
		fprintf('N=%d, kk=%d\n',N,kk);
		tA=tic;
		[avg,stdev,scores]=do_calibration2(curve_len,num_trials,N);
		F=fopen(fname,'w');
		fprintf(F,'%d\n',curve_len);
		fprintf(F,'%d\n',num_trials);
		for j=1:curve_len
			fprintf(F,'%.4f,%.4f\n',avg(j),stdev(j));
		end;
		for j=1:num_trials
			fprintf(F,'%.4f\n',scores(j));
		end;
		fclose(F);
		elapsed=toc(tA);
		%fprintf('Elapsed: %f\n',elapsed);
		est_time_factor=elapsed/N;
		est_time=0;
		for kk2=kk:700
			est_time=est_time+incr^kk2*est_time_factor;
		end;
		fprintf('Estimated #hours remaining: %f\n',est_time/60/60);
	end;
	kk=kk+1;
end;

end

function [avg,stdev,scores]=do_calibration2(curve_len,num_trials,N)

A=zeros(curve_len,num_trials);
for trial=1:num_trials
	X=rand(1,N);
	opts.minsize=4;
	curve=get_isosplit_curve(X,curve_len,opts);
	A(:,trial)=curve;
end;
Asorted=sort(A,2);
avg=squeeze(mean(A,2)); %curve_len x 1
stdev=squeeze(sqrt(var(A,[],2))); %curve_len x 1
scores=zeros(num_trials,1); %num_trials x 1
inds=find(stdev~=0);
if (length(inds)>0)
	scores(:)=max((repmat(avg(inds,1),1,num_trials)-A(inds,:))./repmat(stdev(inds,1),1,num_trials),[],1);
end;
scores=sort(scores,1);

end

function test_isosplit1d 

close all;
N=100000;
for ct=1:15
	X=[randn(1,N/2),randn(1,N/2)+ct/3];
	opts.verbose=0;
	tA=tic;
	[labels,pp]=isosplit1d(X,opts);
	toc(tA)

	if (pp<0.9) labels=ones(size(labels)); end;
	colors='rgbkymc';
	figure;
	[~,bins0]=hist(X,length(X)/3);
	num_labels=max(labels);
	if (pp<0.9) labels=ones(size(labels)); end;
	for j=1:num_labels
		[v0,b0]=hist(X(labels==j),bins0);
		h=bar(b0,v0);
		col=colors(mod(j-1,length(colors))+1);
		set(h,'FaceColor',col,'EdgeColor',col);
		if (j==1) hold on; end;
	end;
	title(sprintf('pp = %.2f\n',pp));
	pause(0.1);
end;

end

File: isosplit/isosplit1d.cpp
#include <QCoreApplication>
#include "isosplit1d.h"
#include <QVector>
#include "mda.h"
#include <math.h>
#include "jisotonic.h"
#include <QDir>
#include <QMap>
#include <stdio.h>
#include <QStringList>
#include <QDebug>
#include <QTime>

void jisotonic_updown(int N,double *out,double *in,double *weights) {
	double *B1=(double *)malloc(sizeof(double)*N);
	double *MSE1=(double *)malloc(sizeof(double)*N);
	double *B2=(double *)malloc(sizeof(double)*N);
	double *MSE2=(double *)malloc(sizeof(double)*N);
	double *in_reversed=(double *)malloc(sizeof(double)*N);
    double *weights_reversed=0;

    for (int j=0; j<N; j++) {in_reversed[j]=in[N-1-j];}
	if (weights) {
        weights_reversed=(double *)malloc(sizeof(double)*N);
        for (int j=0; j<N; j++) {weights_reversed[j]=weights[N-1-j];}
	}
	jisotonic(N,B1,MSE1,in,weights);
	jisotonic(N,B2,MSE2,in_reversed,weights_reversed);
	for (int j=0; j<N; j++) MSE1[j]+=MSE2[N-1-j];
	double bestval=MSE1[0];
	int best_ind=0;
	for (int j=0; j<N; j++) {
		if (MSE1[j]<bestval) {
			bestval=MSE1[j];
			best_ind=j;
		}
	}
	jisotonic(best_ind+1,B1,MSE1,in,weights);
	jisotonic(N-best_ind,B2,MSE2,in_reversed,weights_reversed);
	for (int j=0; j<=best_ind; j++) out[j]=B1[j];
	for (int j=0; j<N-best_ind-1; j++) out[N-1-j]=B2[j];

	free(B1);
	free(MSE1);
	free(B2);
	free(MSE2);
	free(in_reversed);
	if (weights_reversed) free(weights_reversed);
}

void jisotonic_downup(int N,double *out,double *in,double *weights) {
	double *in_neg=(double *)malloc(sizeof(double)*N);

	for (int j=0; j<N; j++) in_neg[j]=-in[j];
	jisotonic_updown(N,out,in_neg,weights);
	for (int j=0; j<N; j++) out[j]=-out[j];

	free(in_neg);
}

void quick_sort (double *a, int n) {
	int i, j;
	double p,t;
	if (n < 2)
		return;
	p = a[n / 2];
	for (i = 0, j = n - 1;; i++, j--) {
		while (a[i] < p)
			i++;
		while (p < a[j])
			j--;
		if (i >= j)
			break;
		t = a[i];
		a[i] = a[j];
		a[j] = t;
	}
	quick_sort(a, i);
	quick_sort(&a[i], n - i);
}

void sort(int N,double *out,double *in) {
	for (int i=0; i<N; i++) out[i]=in[i];
	quick_sort(out,N);
//	QVector<double> in0(N);
//	for (int j=0; j<N; j++) in0[j]=in[j];
//	qSort(in0);
//	for (int j=0; j<N; j++) out[j]=in0[j];
}

bool get_isosplit_curve(int curve_len,int N,double *curve,double &cutpoint,double *X_in,int minsize) {
	QTime timer; timer.start();
	if (N<minsize*2) {
		for (int jj=0; jj<curve_len; jj++) curve[jj]=0;
		cutpoint=0;
		return true;
	}
    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();
    Mda X_mda; X_mda.allocate(1,N); double *X=X_mda.dataPtr();
    sort(N,X,X_in);
    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();
	Mda spacings_mda; spacings_mda.allocate(1,N); double *spacings=spacings_mda.dataPtr();
	spacings[0]=X[1]-X[0];
	spacings[N-1]=X[N-1]-X[N-2];
	for (int j=1; j<N-1; j++) {
		spacings[j]=(X[j+1]-X[j-1])/2;
		if (spacings[j]<=0) {
            printf ("ERROR: in get_isosplit_curve, spacings must be strictly positive -- did you forget to sort, or are the datapoints not distinct? spacings[%d]=%g\n",j,spacings[j]);
			return false;
		}
	}
    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();
	Mda logdensity_mda; logdensity_mda.allocate(1,N); double *logdensity=logdensity_mda.dataPtr();
	Mda sqrt_spacings_mda; sqrt_spacings_mda.allocate(1,N); double *sqrt_spacings=sqrt_spacings_mda.dataPtr();
    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();
	for (int j=0; j<N; j++) {
		logdensity[j]=log(1/spacings[j]);
		sqrt_spacings[j]=sqrt(spacings[j]);
	}
    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();
	Mda fit1_mda; fit1_mda.allocate(1,N); double *fit1=fit1_mda.dataPtr();
	Mda resid1_mda; resid1_mda.allocate(1,N); double *resid1=resid1_mda.dataPtr();
	Mda fit2_mda; fit2_mda.allocate(1,N); double *fit2=fit2_mda.dataPtr();
	Mda fit2_sorted_mda; fit2_sorted_mda.allocate(1,N); double *fit2_sorted=fit2_sorted_mda.dataPtr();
    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();

	jisotonic_updown(N,fit1,logdensity,0);

    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();
	for (int j=0; j<N; j++) resid1[j]=logdensity[j]-fit1[j];
	jisotonic_downup(N,fit2,resid1,sqrt_spacings);
    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();
	double mean_fit2=0;
	for (int j=0; j<N; j++) {
		mean_fit2+=fit2[j];
	}
	mean_fit2/=N;
	for (int j=0; j<N; j++) {
		fit2[j]-=mean_fit2;
	}
	fit2=&fit2[minsize-1];
	N=N-2*minsize+2;
    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();
	sort(N,fit2_sorted,fit2);
    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();
	double sum0=0;
	for (int k=0; k<curve_len; k++) {
		if (k<curve_len) {
			sum0+=fit2_sorted[k];
			curve[k]=sum0/(k+1);
		}
		else curve[k]=0;
	}
	double minval=fit2[0];
	int minind=0;
	for (int j=0; j<N; j++) {
		if (fit2[j]<minval) {
			minval=fit2[j];
			minind=j;
		}
	}
    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();
	minind+=minsize-1;
	if (minind==0) {cutpoint=X[0]; return true;}
	if (minind==N-1) {cutpoint=X[N-1]; return true;}
	if (spacings[minind-1]>spacings[minind+1]) {
		cutpoint=(X[minind-1]+X[minind])/2;
		return true;
	}
	else {
		cutpoint=(X[minind]+X[minind+1])/2;
		return true;
	}
}

struct calibration_file {
	int curve_len;
	int num_trials;
	QVector<double> avg;
	QVector<double> stdev;
	QVector<double> scores;
};
static QMap<int,calibration_file> s_calibration_files;
static QString s_calibration_dir;
static QList<int> s_calibration_Ns;


QString find_directory_path(QString base_path,QString dirname) {
	QStringList folders=QDir(base_path).entryList(QStringList("*"),QDir::Dirs|QDir::NoDotAndDotDot,QDir::Name);
	if (folders.contains(dirname)) return base_path+"/"+dirname;
	for (int i=0; i<folders.count(); i++) {
		QString tmp=find_directory_path(base_path+"/"+folders[i],dirname);
		if (!tmp.isEmpty()) return tmp;
	}
	return "";
}

void set_calibration_dir() {
	QString base_path=qApp->applicationDirPath()+"/..";
	QString path=find_directory_path(base_path,"isosplit_calibration");
	if (path.isEmpty()) return;
	s_calibration_dir=path;
}

void set_calibration_Ns() {
	s_calibration_Ns.clear();
	if (s_calibration_dir.isEmpty()) return;
	QStringList list=QDir(s_calibration_dir).entryList(QStringList("*"),QDir::Files,QDir::Name);
	for (int i=0; i<list.count(); i++) {
		QString fname=list[i];
		int ind1=fname.lastIndexOf("_");
		int ind2=fname.lastIndexOf(".");
		if ((ind1>=0)&&(ind2>=ind1+2)) {
			int N0=fname.mid(ind1+1,ind2-ind1-1).toInt();
			if (N0>0) s_calibration_Ns << N0;
		}
	}
	qSort(s_calibration_Ns);
}

void load_calibration_file(int N) {
	QString path=QString("%1/isosplit_calibration_%2.txt").arg(s_calibration_dir).arg(N);
	QFile FF(path);
	if (FF.open(QFile::ReadOnly|QFile::Text)) {
		QString txt=QString(FF.readAll());
		FF.close();
		QList<QString> lines=txt.split("\n");
		if (lines.count()>10) {
			calibration_file CF;
			CF.curve_len=lines[0].toInt();
			CF.num_trials=lines[1].toInt();
			for (int j=0; j<CF.curve_len; j++) {
				QString line=lines.value(2+j);
				QList<QString> tmp=line.split(",");
				if (tmp.count()!=2) return;
				CF.avg << tmp.value(0).toDouble();
				CF.stdev << tmp.value(1).toDouble();
			}
			for (int j=0; j<CF.num_trials; j++) {
				QString line=lines.value(2+CF.curve_len+j);
				CF.scores << line.toDouble();
			}
			s_calibration_files[N]=CF;
		}
		else return;
	}

}

bool read_isosplit_calibration(int N,QVector<double> &avg,QVector<double> &stdev,QVector<double> &scores) {
	if (s_calibration_dir.isEmpty()) set_calibration_dir();
    if (s_calibration_dir.isEmpty()) {printf ("Error: unable to find isosplit calibration directory (isosplit_calibration)\n"); return false;}
	if (s_calibration_Ns.isEmpty()) set_calibration_Ns();
	int ii=-1;
	while ((ii+2<s_calibration_Ns.count())&&(s_calibration_Ns[ii+1]<=N)) ii++;
    if ((ii<0)||(ii+1>=s_calibration_Ns.count())) {printf ("Error: calibration out of range (N=%d)\n",N); return false;}
	int N1=s_calibration_Ns[ii];
	int N2=s_calibration_Ns[ii+1];
	if (N1==N) N2=N1;
	if (!s_calibration_files.contains(N1)) {
		load_calibration_file(N1);
        if (!s_calibration_files.contains(N1)) {printf ("Error: unable to find calibration file (N=%d)\n",N1); return false;}
	}
	if (!s_calibration_files.contains(N2)) {
		load_calibration_file(N2);
        if (!s_calibration_files.contains(N2)) {printf ("Error: unable to find calibration file (N=%d)\n",N2); return false;}
	}
	if (N1==N2) {
		calibration_file *CF=&s_calibration_files[N1];
		avg=CF->avg; stdev=CF->stdev; scores=CF->scores;
	}
	else {
		double pct=(N-N1)*1.0/(N2-N1);
		calibration_file *CF1=&s_calibration_files[N1];
		calibration_file *CF2=&s_calibration_files[N2];
		avg=CF1->avg; stdev=CF1->stdev; scores=CF1->scores;
		for (int i=0; i<avg.count(); i++) {
			avg[i]=avg[i]+pct*(CF2->avg[i]-avg[i]);
			stdev[i]=stdev[i]+pct*(CF2->stdev[i]-stdev[i]);
		}
		for (int i=0; i<scores.count(); i++) {
			scores[i]=scores[i]+pct*(CF2->scores[i]-scores[i]);
		}
	}
	return true;
}

bool isosplit1d(int N,int *labels,double &pp,double *X) {
	QTime timer; timer.start();
	int minsize=4; //hard coded to be consistent with calibration

    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();
	QVector<double> avg(N),stdev(N),scores(N);
	if (!read_isosplit_calibration(N,avg,stdev,scores)) {
        printf ("ERROR: problem in read_isosplit_calibration.\n");
		return false;
	}
	int curve_len=avg.count();
	if (curve_len==0) {
        printf ("ERROR: calibration not performed for N=%d\n",N);
		return false;
	}
	int num_trials=scores.count();
    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();

	//compute the curve and cutpoint
	Mda curve_mda; curve_mda.allocate(1,curve_len); double *curve=curve_mda.dataPtr();
	double cutpoint;
    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();
	if (!get_isosplit_curve(curve_len,N,curve,cutpoint,X,minsize)) {
        printf ("ERROR in get_isosplit_curve\n");
		return false;
	}
    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();

	//compute the score by comparing to calibration avg and stdev and then compute the pp
	double score0=0;
	for (int ii=0; ii<curve_len; ii++) {
		if (stdev[ii]) {
			double diff0=avg[ii]-curve[ii];
			//we need a change of at least 0.1 to count it
			//this is important in the case where stdev is extremely close to 0
			if (qAbs(diff0)<0.1) diff0=0;
			double tmp_score=diff0/stdev[ii];
			if (tmp_score>score0) score0=tmp_score;
		}
	}
	if (score0!=0) {
		double tmp_ct=0;
		for (int jj=0; jj<scores.count(); jj++) {
			if (scores[jj]<score0) tmp_ct++;
		}
		pp=tmp_ct/num_trials;
	}
	else {
		pp=0;
	}
    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();

	//set the labels
	for (int kk=0; kk<N; kk++) {
		if (X[kk]<cutpoint) labels[kk]=1;
		else labels[kk]=2;
	}
    //qDebug() << "ELAPSED" << __FILE__ << __LINE__ << timer.elapsed(); timer.start();

	return true;
}
